File: packages/payment/android/src/main/java/com/getcapacitor/community/stripe/paymentsheet/PaymentSheetExecutor.java
package com.getcapacitor.community.stripe.paymentsheet;

import android.app.Activity;
import android.content.Context;

import androidx.core.util.Supplier;

import com.getcapacitor.Bridge;
import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;
import com.getcapacitor.community.stripe.models.Executor;
import com.google.android.gms.common.util.BiConsumer;
import com.stripe.android.paymentsheet.PaymentSheet;
import com.stripe.android.paymentsheet.PaymentSheetResult;

public class PaymentSheetExecutor extends Executor {

    public PaymentSheet paymentSheet;
    private final JSObject emptyObject = new JSObject();
    private PaymentSheet.Configuration paymentConfiguration;

    private String paymentIntentClientSecret;
    private String setupIntentClientSecret;

    public PaymentSheetExecutor(
        Supplier<Context> contextSupplier,
        Supplier<Activity> activitySupplier,
        BiConsumer<String, JSObject> notifyListenersFunction,
        String pluginLogTag
    ) {
        super(contextSupplier, activitySupplier, notifyListenersFunction, pluginLogTag, "PaymentSheetExecutor");
        this.contextSupplier = contextSupplier;
    }

    public void createPaymentSheet(final PluginCall call) {
        paymentIntentClientSecret = call.getString("paymentIntentClientSecret", null);
        setupIntentClientSecret = call.getString("setupIntentClientSecret", null);

        String customerEphemeralKeySecret = call.getString("customerEphemeralKeySecret", null);
        String customerId = call.getString("customerId", null);

        if (paymentIntentClientSecret == null && setupIntentClientSecret == null) {
            String errorText = "Invalid Params. This method require paymentIntentClientSecret or setupIntentClientSecret.";
            notifyListenersFunction.accept(PaymentSheetEvents.FailedToLoad.getWebEventName(), new JSObject().put("error", errorText));
            call.reject(errorText);
            return;
        }

        if (customerId != null && customerEphemeralKeySecret == null) {
            String errorText = "Invalid Params. When you set customerId, you must set customerEphemeralKeySecret.";
            notifyListenersFunction.accept(PaymentSheetEvents.FailedToLoad.getWebEventName(), new JSObject().put("error", errorText));
            call.reject(errorText);
            return;
        }

        String merchantDisplayName = call.getString("merchantDisplayName");

        if (merchantDisplayName == null) {
            merchantDisplayName = "";
        }

        Boolean enableGooglePay = call.getBoolean("enableGooglePay", false);

        final PaymentSheet.CustomerConfiguration customer = customerId != null
            ? new PaymentSheet.CustomerConfiguration(customerId, customerEphemeralKeySecret)
            : null;

        PaymentSheet.BillingDetailsCollectionConfiguration billingDetailsCollectionConfiguration = null;
        JSObject bdCollectionConfiguration = call.getObject("billingDetailsCollectionConfiguration", null);
        if (bdCollectionConfiguration != null){
            String emailCollectionMode = bdCollectionConfiguration.getString("email");
            String nameCollectionMode = bdCollectionConfiguration.getString("name");
            String phoneCollectionMode = bdCollectionConfiguration.getString("phone");
            String addressCollectionMode = bdCollectionConfiguration.getString("address");
            billingDetailsCollectionConfiguration = new PaymentSheet.BillingDetailsCollectionConfiguration(
                    (nameCollectionMode != null && nameCollectionMode.equals("always")) ? PaymentSheet.BillingDetailsCollectionConfiguration.CollectionMode.Always : PaymentSheet.BillingDetailsCollectionConfiguration.CollectionMode.Automatic,
                    (phoneCollectionMode != null && phoneCollectionMode.equals("always")) ? PaymentSheet.BillingDetailsCollectionConfiguration.CollectionMode.Always : PaymentSheet.BillingDetailsCollectionConfiguration.CollectionMode.Automatic,
                    (emailCollectionMode != null && emailCollectionMode.equals("always")) ? PaymentSheet.BillingDetailsCollectionConfiguration.CollectionMode.Always : PaymentSheet.BillingDetailsCollectionConfiguration.CollectionMode.Automatic,
                    (addressCollectionMode != null && addressCollectionMode.equals("full")) ? PaymentSheet.BillingDetailsCollectionConfiguration.AddressCollectionMode.Full : PaymentSheet.BillingDetailsCollectionConfiguration.AddressCollectionMode.Automatic,
                    false
            );
        }

        if (!enableGooglePay) {
            if (bdCollectionConfiguration != null) {
                paymentConfiguration = new PaymentSheet.Configuration.Builder(merchantDisplayName)
                        .customer(customer)
                        .billingDetailsCollectionConfiguration(billingDetailsCollectionConfiguration)
                        .build();
            } else {
                paymentConfiguration = new PaymentSheet.Configuration.Builder(merchantDisplayName)
                        .customer(customer)
                        .build();
            }
        } else {
            Boolean GooglePayEnvironment = call.getBoolean("GooglePayIsTesting", false);

            PaymentSheet.GooglePayConfiguration.Environment environment = PaymentSheet.GooglePayConfiguration.Environment.Production;

            if (GooglePayEnvironment) {
                environment = PaymentSheet.GooglePayConfiguration.Environment.Test;
            }

            if (bdCollectionConfiguration != null) {
                paymentConfiguration = new PaymentSheet.Configuration.Builder(merchantDisplayName)
                        .customer(customer)
                        .googlePay(new PaymentSheet.GooglePayConfiguration(environment, call.getString("countryCode", "US")))
                        .billingDetailsCollectionConfiguration(billingDetailsCollectionConfiguration)
                        .build();
            } else {
                paymentConfiguration = new PaymentSheet.Configuration.Builder(merchantDisplayName)
                        .customer(customer)
                        .googlePay(new PaymentSheet.GooglePayConfiguration(environment, call.getString("countryCode", "US")))
                        .build();
            }

        }

        notifyListenersFunction.accept(PaymentSheetEvents.Loaded.getWebEventName(), emptyObject);
        call.resolve();
    }

    public void presentPaymentSheet(final PluginCall call) {
        try {
            if (paymentIntentClientSecret != null) {
                paymentSheet.presentWithPaymentIntent(paymentIntentClientSecret, paymentConfiguration);
            } else {
                paymentSheet.presentWithSetupIntent(setupIntentClientSecret, paymentConfiguration);
            }
        } catch (Exception ex) {
            call.reject(ex.getLocalizedMessage(), ex);
        }
    }

    public void onPaymentSheetResult(Bridge bridge, String callbackId, final PaymentSheetResult paymentSheetResult) {
        PluginCall call = bridge.getSavedCall(callbackId);

        if (paymentSheetResult instanceof PaymentSheetResult.Canceled) {
            notifyListenersFunction.accept(PaymentSheetEvents.Canceled.getWebEventName(), emptyObject);
            call.resolve(new JSObject().put("paymentResult", PaymentSheetEvents.Canceled.getWebEventName()));
        } else if (paymentSheetResult instanceof PaymentSheetResult.Failed) {
            notifyListenersFunction.accept(
                PaymentSheetEvents.Failed.getWebEventName(),
                new JSObject().put("message", ((PaymentSheetResult.Failed) paymentSheetResult).getError().getLocalizedMessage())
            );
            notifyListenersFunction.accept(PaymentSheetEvents.Failed.getWebEventName(), emptyObject);
            call.resolve(new JSObject().put("paymentResult", PaymentSheetEvents.Failed.getWebEventName()));
        } else if (paymentSheetResult instanceof PaymentSheetResult.Completed) {
            notifyListenersFunction.accept(PaymentSheetEvents.Completed.getWebEventName(), emptyObject);
            call.resolve(new JSObject().put("paymentResult", PaymentSheetEvents.Completed.getWebEventName()));
        }
    }
}

File: packages/payment/android/src/main/java/com/getcapacitor/community/stripe/paymentsheet/PaymentSheetExecutor.kt
package com.getcapacitor.community.stripe.paymentsheet

import android.app.Activity
import android.content.Context
import androidx.core.util.Supplier
import com.getcapacitor.Bridge
import com.getcapacitor.JSObject
import com.getcapacitor.PluginCall
import com.getcapacitor.community.stripe.helper.PaymentSheetHelper
import com.getcapacitor.community.stripe.models.Executor
import com.google.android.gms.common.util.BiConsumer
import com.stripe.android.paymentsheet.PaymentSheet
import com.stripe.android.paymentsheet.PaymentSheet.BillingDetailsCollectionConfiguration.AddressCollectionMode
import com.stripe.android.paymentsheet.PaymentSheet.BillingDetailsCollectionConfiguration.CollectionMode
import com.stripe.android.paymentsheet.PaymentSheetResult
import com.stripe.android.paymentsheet.addresselement.AddressDetails

class PaymentSheetExecutor(
    contextSupplier: Supplier<Context>,
    activitySupplier: Supplier<Activity>,
    notifyListenersFunction: BiConsumer<String, JSObject>,
    pluginLogTag: String
) : Executor(
    contextSupplier,
    activitySupplier,
    notifyListenersFunction,
    pluginLogTag,
    "PaymentSheetExecutor"
) {
    var paymentSheet: PaymentSheet? = null
    private val emptyObject = JSObject()
    private var paymentConfiguration: PaymentSheet.Configuration? = null

    private var paymentIntentClientSecret: String? = null
    private var setupIntentClientSecret: String? = null

    init {
        this.contextSupplier = contextSupplier
    }

    fun createPaymentSheet(call: PluginCall) {
        paymentIntentClientSecret = call.getString("paymentIntentClientSecret", null)
        setupIntentClientSecret = call.getString("setupIntentClientSecret", null)

        val customerEphemeralKeySecret = call.getString("customerEphemeralKeySecret", null)
        val customerId = call.getString("customerId", null)

        val paymentMethodLayout: PaymentSheet.PaymentMethodLayout = when (call.getString("paymentMethodLayout", "automatic")) {
            "horizontal" -> PaymentSheet.PaymentMethodLayout.Horizontal
            "vertical"   -> PaymentSheet.PaymentMethodLayout.Vertical
            "automatic"  -> PaymentSheet.PaymentMethodLayout.Automatic
            else         -> PaymentSheet.PaymentMethodLayout.Automatic
        }

        if (paymentIntentClientSecret == null && setupIntentClientSecret == null) {
            val errorText =
                "Invalid Params. This method require paymentIntentClientSecret or setupIntentClientSecret."
            notifyListenersFunction.accept(
                PaymentSheetEvents.FailedToLoad.webEventName,
                JSObject().put("error", errorText)
            )
            call.reject(errorText)
            return
        }

        if (customerId != null && customerEphemeralKeySecret == null) {
            val errorText =
                "Invalid Params. When you set customerId, you must set customerEphemeralKeySecret."
            notifyListenersFunction.accept(
                PaymentSheetEvents.FailedToLoad.webEventName,
                JSObject().put("error", errorText)
            )
            call.reject(errorText)
            return
        }

        var merchantDisplayName = call.getString("merchantDisplayName")

        if (merchantDisplayName == null) {
            merchantDisplayName = ""
        }

        val enableGooglePay = call.getBoolean("enableGooglePay", false)

        val customer: PaymentSheet.CustomerConfiguration? = if (customerId != null
        ) PaymentSheet.CustomerConfiguration(customerId, customerEphemeralKeySecret!!)
        else null


        val defaultBillingDetailsConfiguration = PaymentSheetHelper().fromJSObjectToBillingDetails(call.getObject("defaultBillingDetails", null))
        val shippingDetailsConfiguration = PaymentSheetHelper().fromJSObjectToShippingDetails(call.getObject("shippingDetails", null));
        val billingDetailsCollectionConfiguration = PaymentSheetHelper().fromJSObjectToBillingCollectionConfig(
            call.getObject("billingDetailsCollectionConfiguration", null)
        )

        if (!enableGooglePay!!) {
            paymentConfiguration = PaymentSheet.Configuration.Builder(merchantDisplayName = merchantDisplayName)
                .customer(customer)
                .defaultBillingDetails(defaultBillingDetailsConfiguration)
                .shippingDetails(shippingDetailsConfiguration)
                .billingDetailsCollectionConfiguration(billingDetailsCollectionConfiguration)
                .paymentMethodLayout(paymentMethodLayout)
                .build()
        } else {
            val googlePayEnvironment = call.getBoolean("GooglePayIsTesting", false)

            var environment: PaymentSheet.GooglePayConfiguration.Environment =
                PaymentSheet.GooglePayConfiguration.Environment.Production

            if (googlePayEnvironment!!) {
                environment = PaymentSheet.GooglePayConfiguration.Environment.Test
            }

            paymentConfiguration = PaymentSheet.Configuration.Builder(merchantDisplayName = merchantDisplayName)
                .customer(customer)
                .defaultBillingDetails(defaultBillingDetailsConfiguration)
                .shippingDetails(shippingDetailsConfiguration)
                .billingDetailsCollectionConfiguration(billingDetailsCollectionConfiguration)
                .paymentMethodLayout(paymentMethodLayout)
                .googlePay(PaymentSheet.GooglePayConfiguration(
                    environment,
                    call.getString("countryCode", "US")!!,
                    call.getString("currencyCode", null)
                ))
                .build()
        }

        notifyListenersFunction.accept(PaymentSheetEvents.Loaded.webEventName, emptyObject)
        call.resolve()
    }

    fun presentPaymentSheet(call: PluginCall) {
        try {
            if (paymentIntentClientSecret != null) {
                paymentSheet!!.presentWithPaymentIntent(
                    paymentIntentClientSecret!!,
                    paymentConfiguration
                )
            } else {
                paymentSheet!!.presentWithSetupIntent(
                    setupIntentClientSecret!!,
                    paymentConfiguration
                )
            }
        } catch (ex: Exception) {
            call.reject(ex.localizedMessage, ex)
        }
    }

    fun onPaymentSheetResult(
        bridge: Bridge,
        callbackId: String?,
        paymentSheetResult: PaymentSheetResult
    ) {
        val call = bridge.getSavedCall(callbackId)

        if (paymentSheetResult is PaymentSheetResult.Canceled) {
            notifyListenersFunction.accept(PaymentSheetEvents.Canceled.webEventName, emptyObject)
            call.resolve(JSObject().put("paymentResult", PaymentSheetEvents.Canceled.webEventName))
        } else if (paymentSheetResult is PaymentSheetResult.Failed) {
            notifyListenersFunction.accept(
                PaymentSheetEvents.Failed.webEventName,
                JSObject().put(
                    "message",
                    (paymentSheetResult).error.localizedMessage
                )
            )
            notifyListenersFunction.accept(PaymentSheetEvents.Failed.webEventName, emptyObject)
            call.resolve(JSObject().put("paymentResult", PaymentSheetEvents.Failed.webEventName))
        } else if (paymentSheetResult is PaymentSheetResult.Completed) {
            notifyListenersFunction.accept(PaymentSheetEvents.Completed.webEventName, emptyObject)
            call.resolve(JSObject().put("paymentResult", PaymentSheetEvents.Completed.webEventName))
        }
    }
}
